File: typescript/src/useCase/OrderApprovalUseCase.ts
import OrderRepository from "../repository/OrderRepository";
import OrderApprovalRequest from "./OrderApprovalRequest";
import ApprovedOrderCannotBeRejectedError from "./ApprovedOrderCannotBeRejectedError";
import ShippedOrdersCannotBeChangedError from "./ShippedOrdersCannotBeChangedError";
import RejectedOrderCannotBeApprovedError from "./RejectedOrderCannotBeApprovedError";
import OrderStatus from "../domain/OrderStatus";

export default class OrderApprovalUseCase {
    private readonly _orderRepository: OrderRepository;

    constructor(orderRepository: OrderRepository) {
        this._orderRepository = orderRepository;
    }

    public run(request: OrderApprovalRequest): void {
        const order = this._orderRepository.getById(request.orderId);
        if (order.status === OrderStatus.SHIPPED) {
            throw new ShippedOrdersCannotBeChangedError();
        }

        if (request.approved && order.status === OrderStatus.REJECTED) {
            throw new RejectedOrderCannotBeApprovedError();
        }

        if (!request.approved && order.status === OrderStatus.APPROVED) {
            throw new ApprovedOrderCannotBeRejectedError();
        }

        order.status = request.approved ? OrderStatus.APPROVED : OrderStatus.REJECTED;
        this._orderRepository.save(order);
    }
}

File: java/src/main/java/it/gabrieletondi/telldontaskkata/useCase/OrderApprovalUseCase.java
package it.gabrieletondi.telldontaskkata.useCase;

import it.gabrieletondi.telldontaskkata.domain.Order;
import it.gabrieletondi.telldontaskkata.domain.OrderStatus;
import it.gabrieletondi.telldontaskkata.repository.OrderRepository;

public class OrderApprovalUseCase {
    private final OrderRepository orderRepository;

    public OrderApprovalUseCase(OrderRepository orderRepository) {
        this.orderRepository = orderRepository;
    }

    public void run(OrderApprovalRequest request) {
        final Order order = orderRepository.getById(request.getOrderId());
        boolean isApproved = request.isApproved();
        if (order.getStatus().equals(OrderStatus.SHIPPED)) {
            throw new ShippedOrdersCannotBeChangedException();
        }
        if (isApproved) {
            if (order.getStatus().equals(OrderStatus.REJECTED)) {
                throw new RejectedOrderCannotBeApprovedException();
            }

            order.setStatus(OrderStatus.APPROVED);
        } else {
            if (order.getStatus().equals(OrderStatus.APPROVED)) {
                throw new ApprovedOrderCannotBeRejectedException();
            }
            order.setStatus(OrderStatus.REJECTED);
        }
        orderRepository.save(order);
    }
}
